File: downloads/code/LeetCode/Java/Evaluate-Reverse-Polish-Notation.java
public class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> nums = new Stack<>();
        int a,b;
        for (String s : tokens){
            if (s.equals("+")){
                nums.push(nums.pop() + nums.pop());
            }
            else if (s.equals("*")){
                nums.push(nums.pop() * nums.pop());
            }
            else if (s.equals("/")){
                b = nums.pop();
                a = nums.pop();
                nums.push(a/b);
            }
            else if (s.equals("-")){
                b = nums.pop();
                a = nums.pop();
                nums.push(a-b);
            }
            else 
                nums.push(Integer.parseInt(s));
        
        }
        return nums.pop();
    }
}

File: downloads/code/LeetCode/Python/Evaluate-Reverse-Polish-Notation.py
class Solution:
    # @param tokens, a list of string
    # @return an integer
    def evalRPN(self, tokens):
        stack = []
        for i in tokens:
            if i not in ('+', '-', '*', '/'):
                stack.append(int(i))
            else:
                # be careful of the order of op1 and op2
                op2 = stack.pop()
                op1 = stack.pop()
                if i == '+': stack.append(op1 + op2)
                elif i == '-': stack.append(op1 - op2)
                elif i == '*': stack.append(op1 * op2)
                else: stack.append(int(op1 * 1.0 / op2))
        return stack[0]
